src/extension.ts

```diff
@@ -0,0 +1,152 @@
+
+			if (pubspecContainsOverReactFormat && !overReactFormatRangeIsValid) {
+				this.showChannelMessage(`OverReact Format range is not acceptable. Bump the minimum to 3.1.0 to use OverReact Format on Save.`);
+			}
+			this.useOverReactFormat =  pubspecContainsOverReactFormat && overReactFormatRangeIsValid;
+		}
+		// No else condition because there's no penalty for the project not being a Dart project.
+		// The `onDocumentSave` command will just be short-circuited if it is run on non-Dart files.
+	}
+
+	buildCommand(fileName: string) {
+		let command:string;
+
+		const customLineLength = this.config.get<Number>('customLineLength', 0);
+		const shouldDetectLineLength = this.config.get<Boolean>('detectCustomLineLength');
+		const shouldUseCustomLineLength = customLineLength > 0;
+
+		if (shouldUseCustomLineLength && shouldDetectLineLength) {
+			this.showChannelMessage(`Both a custom line-length value and detectCustomLineLength set to true. Skipping line-length detection.`);
+		}
+
+		if (this.useOverReactFormat) {
+			if (shouldUseCustomLineLength) {
+				command =  `pub run over_react_format ${fileName} -l ${customLineLength}`;
+			} else {
+				const detectLineLengthFlag = shouldDetectLineLength && !shouldUseCustomLineLength ? "--detect-line-length" : "";
+				command =  `pub run over_react_format ${fileName} -p ${this.projectDir} ${detectLineLengthFlag}`;
+			} 
+		} else {
+			// TODO add logic to detect line-length from dart_dev's config.dart
+			command = `dartfmt -w ${fileName} ${shouldUseCustomLineLength ? `-l ${customLineLength}` : ''}`;
+		}
+
+		return command;
+	}
+
+	loadConfig() {
+		this.config = vscode.workspace.getConfiguration('overReact.formatOnSave');
+	}
+	
+	private showEnablingChannelMessage () {
+		const message = `Run OverReact Format on Save is ${this.getEnabled() ? 'enabled' : 'disabled'}`;
+		this.showChannelMessage(message);
+		this.showStatusMessage(message);
+	}
+
+	private showChannelMessage(message: string) {
+		this.channel.appendLine(message);
+	}
+
+	getEnabled(): boolean {
+		return !!this.context.globalState.get('enabled', true);
+	}
+		
+	setEnabled(enabled: boolean) {
+		this.context.globalState.update('enabled', enabled);
+		this.showEnablingChannelMessage();
+	}
+
+	private showStatusMessage(message: string) {
+		const disposable = vscode.window.setStatusBarMessage(message, 3000);
+		this.context.subscriptions.push(disposable);
+	}
+
+	onDocumentSave(document: vscode.TextDocument) {
+		if (!this.getEnabled() || document.languageId !== 'dart') {
+			return;
+		}
+
+		this.showChannelMessage(`Running ${this.useOverReactFormat ? 'OverReact Format' : 'dartfmt'}...`);
+
+		const command = this.buildCommand(document.fileName); 
+
+		this.showChannelMessage(command);
+
+		const child:process.ChildProcess = process.exec(command, {cwd: this.projectDir});
```

Comment:
look up `process.exec` -- will it allow multiple commands in a single one? Looks like probably, since it's passed as a string.

Reply:
Yup, probably want `execFile` instead of `exec`. That does the normal thing of accepting the binary to execute and the list of args to run as separate parameters, rather than just flarping a string onto the cli.